[netatmo] Support for favorite weather stations (#8020 #8111)

Novanic opened the pull request:
This pull-request enables users to include weather data of favorite weather stations.

See issues #8020 and #8111 for more information.

lolodomo commented:
Do we really need a bridge configuration setting for that ? IMHO, we don't.

When I implemented this change myself few days ago, I noticed something annoying, the user has nothing to distinguish in the discovery inbox between the favorite stations and his own stattion. Additionally, there is nothing in the inbox to know what module(s) are linked to one particular station. This leads to the conclusion that the naming of things in the inbox has to be improved when this feature is implemented and that's why I did not create a PR immediately.

Novanic commented:
Ah, really, there is also issue #8020 . I haven't seen that. :-(

I think the bridge is required, because a Netatmo account is required to use the API and to mark favorite stations. These information is entered within the bridge configuration ("Netatmo API").

I have also noticed the naming problems, but I thought it's just up to the user which name of the main module is configured. I will check if there is another naming option to solve that. But at least all modules/things of the inbox are linked correctly... ;-)

lolodomo commented:
I was probably not clear  I don't see the need for a configuration setting. Just use true to search own and favorite stations.

lolodomo commented:
Regarding the naming, my initial attempt was to add " (fav(" to the station name. 
Take care, sometimes the module name is null.

Novanic commented:
> I was probably not clear I don't see the need for a configuration setting. Just use true to search own and favorite stations.

Ok, that is right, when the name is more concrete it shouldn't be necessary. Users which don't want to integrate their favorite stations, can just exclude them.

Novanic commented:
> Regarding the naming, my initial attempt was to add " (fav)" to the station name.
> Take care, sometimes the module name is null.

I saw also the street name of the station which should be good to recognize the stations. I will see if it is accessible and useable.

lolodomo commented:
But they can just ignore them in the inbox.

Novanic commented:
> > Regarding the naming, my initial attempt was to add " (fav)" to the station name.
> > Take care, sometimes the module name is null.
> 
> I saw also the street name of the station which should be good to recognize the stations. I will see if it is accessible and useable.
> 
> @lolodomo Should I continue with the issue and pull-request or do you want to continue your already started work? Both would be fine for me.

Hm, the street name is missing in the response of /getstationsdata. But it is available when using endpoint /getpublicdata, but this is slow and complicated to use because a position/place range is required as the input parameters... I think it is a bug within the Netatmo API within the /getstationsdata API that the street name isn't within the output... That makes it complicated / impossible to realize.

Novanic commented:
The default names/labels of the weather station and weather module things are now changed to ...
- let the user recognize which are things from favorite stations (and not from the own station)
- display which modules belong to which station

The rules are the following:
1. When it is a favorite station or module "(favorite)" is added to the thing name/label
2. The type is now added to the name/label, because the station name isn't precise enough (it contains only the name of the city)
3. When a module name is available it is contained within the name/label. When a module name isn't available, the localized module type is within the name/label.
4. The module contains the station name (to make clear which module belongs to which station)

lolodomo commented on bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/discovery/NetatmoModuleDiscoveryService.java:
```diff
@@ -195,4 +202,46 @@ private ThingUID findThingUID(String thingType, String thingId) throws IllegalAr
+    private String createWeatherStationName(NAMain station, boolean isFavorite) {
+        StringBuilder nameBuilder = new StringBuilder();
+        nameBuilder.append(localizeType(station.getType()));
+        if(station.getStationName() != null) {
```
Please apply the formatting standards set for the project. This is something to setup in your IDE. Additionally you may have a command to reformat a source file but I don't know it. 
As an example, there is a space between the if and it's condition.

Novanic replied:
ok, I have now executed the Maven goal spotless:apply. There were a few other files which had formatting errors. Now spotless:check is also running successfully. :-)

lolodomo commented on bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/discovery/NetatmoModuleDiscoveryService.java:
```diff
@@ -195,4 +202,46 @@ private ThingUID findThingUID(String thingType, String thingId) throws IllegalAr
+        return nameBuilder.toString();
+    }
+
+    private String localizeType(String typeName) {
```
I think the core framework already provides a method for that. I have to search.

Novanic replied:
It is described here and I found usages/examples within other bundles: https://www.openhab.org/docs/developer/utils/i18n.html

But I'm also wondering why it is so complicated to translate a String... It would be neat if the core framework would provide a shorter / easier to use solution.

lolodomo commented on bundles/org.openhab.binding.netatmo/src/test/java/org/openhab/binding/netatmo/internal/discovery/NetatmoModuleDiscoveryServiceTest.java:
```diff
@@ -0,0 +1,278 @@
+
+        List<DiscoveryResult> discoveredThings = service.getDiscoveredThings();
+        assertEquals(1, discoveredThings.size());
+        //Expected is the type name + station name, because both is available
```
Both are... 
Space before "Expected"

Novanic replied:
ok, I have now executed the Maven goal spotless:apply. There were a few other files which had formatting errors. Now spotless:check is also running successfully. :-)

Novanic commented:
> LGTM
> 
> Maybe "both are available" rather than "both is available"?

ok, fixed. Thank you

Novanic commented:
Ok, thank you. :-)

Can this get merged soon? 2.5.7 will get built probably soon, it's every month between the 17. and 20. day.

lolodomo commented:
@fwolter : any chance that you review and merge this small enhancement just in time for the 2.5.7 relase ? We are already both to be ok with this change.

lolodomo commented:
Thank you
